### PR TITLE
KOGITO-2812 - add explainability service

### PR DESCRIPTION
**Pull request by r00ta**
Jira: https://issues.redhat.com/browse/KOGITO-2812

The aim of this pr is to add the support of the explainability service in the operator. The architecture looks like the following: 

Screenshot from 2020-08-25 10-32-12

In particular we have 
1) the trusty service that will generate some explainability requests
2) the explainability service takes the requests, perturbs the inputs of the decision, makes an http call to the kogito runtime service to retrieve the results and then use the explainability library to calculate the features importance.
3) the explainability service generates explainability result events that are consumed by the trusty service. 
4) the explainability results are exposed via an endpoint.

In order to reach the kogito runtime service, the route of that service is injected in the explainability service instances. 

Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.MD#sending-a-pull-request)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

### CODE REVIEW

**Comment by r00ta** [comment]
depends on https://github.com/kiegroup/kogito-images/pull/193

File: test/steps/kogitoexplainability.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+type executionsResponse struct {
+	Executions []execution `json:"executions"`
+}
+
+type execution struct {
+	ExecutionID string `json:"executionId"`
+}
+
+func (data *Data) explainabilityResultIsAvailable(timeoutInMin int) error {
```
Comment (r00ta) [inline test/steps/kogitoexplainability.go:73]: This test is not supposed to work at the moment, since the implementations in the trusty and explainability services are still in progress
https://github.com/kiegroup/kogito-runtimes/pull/720
https://github.com/kiegroup/kogito-apps/pull/410

**Comment by r00ta** [comment]
Hi @ricardozanini , @radtriste and others, this pr should be ready to be reviewed (even if the final integration test is not working atm since the implementation on the services side is still in progress). 
Looking forward your feedbacks!

**Comment by ricardozanini** [comment]
> Hi @ricardozanini , @radtriste and others, this pr should be ready to be reviewed (even if the final integration test is not working atm since the implementation on the services side is still in progress).
> Looking forward your feedbacks!

Hi! Thanks for the PR!

I'd rather wait for everything to be working to coordinate the review of this PR. Unit and Bats tests are failing as well.

**Comment by ricardozanini** [comment]
One more thing: please do not use Routes or do not count on external endpoints to make the calls between the services. Use [Kubernetes services](https://kubernetes.io/docs/concepts/services-networking/service/) instead.

**Comment by r00ta** [comment]
> One more thing: please do not use Routes or do not count on external endpoints to make the calls between the services. Use [Kubernetes services](https://kubernetes.io/docs/concepts/services-networking/service/) instead.

Hi @ricardozanini , I've implemented the same approach used for the data-index service and the management console for example. Just for my understanding, Is this a different use case?

**Comment by ricardozanini** [comment]
> > One more thing: please do not use Routes or do not count on external endpoints to make the calls between the services. Use [Kubernetes services](https://kubernetes.io/docs/concepts/services-networking/service/) instead.
> 
> Hi @ricardozanini , I've implemented the same approach used for the data-index service and the management console for example. Just for my understanding, Is this a different use case?

We will remove this integration and yes, it's different. Management console needs to access the external endpoint since it's a javascript app. We will solve this with an NGINX reverse proxy on Management Console side calling the internal services instead.

**Comment by ricardozanini** [comment]
@r00ta please solve bats and unit tests first before running the test on Jenkins side.

**Comment by r00ta** [comment]
> > > One more thing: please do not use Routes or do not count on external endpoints to make the calls between the services. Use [Kubernetes services](https://kubernetes.io/docs/concepts/services-networking/service/) instead.
> > 
> > 
> > Hi @ricardozanini , I've implemented the same approach used for the data-index service and the management console for example. Just for my understanding, Is this a different use case?
> 
> We will remove this integration and yes, it's different. Management console needs to access the external endpoint since it's a javascript app. We will solve this with an NGINX reverse proxy on Management Console side calling the internal services instead.

Thanks for the clarifications/tips @ricardozanini , I'll change it accordingly!

**Comment by r00ta** [comment]
> One more thing: please do not use Routes or do not count on external endpoints to make the calls between the services. Use [Kubernetes services](https://kubernetes.io/docs/concepts/services-networking/service/) instead.

Hi @ricardozanini , I went through the documentation and iiuc we have two options: 
1) Use the environment variables created by kubernetes in every service: for example if we deploy the trusty service and then the expl service, we would find the env variables `TRUSTY_SERVICE_HOST` and `TRUSTY_SERVICE_PORT` that we can use to make some calls. However, this approach works only if the deployment follow a particular order. 
```
When you have a Pod that needs to access a Service, and you are using the environment variable method to publish the port and cluster IP to the client Pods, you must create the Service before the client Pods come into existence. Otherwise, those client Pods won't have their environment variables populated.
``` 

2) use a DNS and perform a simple lookup on the service name. We would have to install https://coredns.io/ . 

While I'm writing this, I'm realizing that unfortunately we don't know the name of the kogito runtime service upfront. And even if we use the approach 1 or 2, the expl service needs to know at least the name of the service; but this is custom. wdyt?

**Comment by r00ta** [comment]
> > One more thing: please do not use Routes or do not count on external endpoints to make the calls between the services. Use [Kubernetes services](https://kubernetes.io/docs/concepts/services-networking/service/) instead.
> 
> Hi @ricardozanini , I went through the documentation and iiuc we have two options:
> 
> 1. Use the environment variables created by kubernetes in every service: for example if we deploy the trusty service and then the expl service, we would find the env variables `TRUSTY_SERVICE_HOST` and `TRUSTY_SERVICE_PORT` that we can use to make some calls. However, this approach works only if the deployment follow a particular order.
> 
> ```
> When you have a Pod that needs to access a Service, and you are using the environment variable method to publish the port and cluster IP to the client Pods, you must create the Service before the client Pods come into existence. Otherwise, those client Pods won't have their environment variables populated.
> ```
> 
> 1. use a DNS and perform a simple lookup on the service name. We would have to install https://coredns.io/ .
> 
> While I'm writing this, I'm realizing that unfortunately we don't know the name of the kogito runtime service upfront. And even if we use the approach 1 or 2, the expl service needs to know at least the name of the service; but this is customized by the user. wdyt?

Sorry @ricardozanini , forget it. I found out we don't need to do that since the url is passed directly in the event that the expl service receives. See https://github.com/kiegroup/kogito-apps/pull/410#discussion_r477138488 . 
I will remove this code

**Comment by ricardozanini** [comment]
@vaibhavjainwiz I believe we could focus our efforts towards the refactoring you proposed a couple weeks ago after finishing the KogitoInfra refactoring. With that we would be able to rip off all those controllers. This service, Jobs Service, TrustyAI, Management Console all looks really the same thing from the operator perspective, with minor details.

We would have one controller and one CR like `KogitoSupportService` => kind `DataIndex` for instance.

**Comment by vaibhavjainwiz** [comment]
> @vaibhavjainwiz I believe we could focus our efforts towards the refactoring you proposed a couple weeks ago after finishing the KogitoInfra refactoring. With that we would be able to rip off all those controllers. This service, Jobs Service, TrustyAI, Management Console all looks really the same thing from the operator perspective, with minor details.
> 
> We would have one controller and one CR like `KogitoSupportService` => kind `DataIndex` for instance.

sure @ricardozanini

File: pkg/apis/app/v1alpha1/kogitoexplainability_types.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+// +k8s:openapi-gen=true
+// +kubebuilder:resource:path=kogitoexplainabilities,scope=Namespaced
+// +kubebuilder:printcolumn:name="Replicas",type="integer",JSONPath=".spec.replicas",description="Number of replicas set for this service"
+// +kubebuilder:printcolumn:name="Image",type="string",JSONPath=".status.image",description="Base image for this service"
+// +kubebuilder:printcolumn:name="Endpoint",type="string",JSONPath=".status.externalURI",description="External URI to access this service"
+// +operator-sdk:gen-csv:customresourcedefinitions.displayName="Kogito Explainability"
+// +operator-sdk:gen-csv:customresourcedefinitions.resources="Deployments,apps/v1"
+// +operator-sdk:gen-csv:customresourcedefinitions.resources="Routes,route.openshift.io/v1"
+// +operator-sdk:gen-csv:customresourcedefinitions.resources="Services,v1"
+// +operator-sdk:gen-csv:customresourcedefinitions.resources="KafkaTopics,kafka.strimzi.io/v1beta1"
```
Comment (ricardozanini) [inline pkg/apis/app/v1alpha1/kogitoexplainability_types.go:55]:
```suggestion
// +operator-sdk:gen-csv:customresourcedefinitions.displayName="Kogito Explainability"
// +operator-sdk:gen-csv:customresourcedefinitions.resources="Deployment,apps/v1"
// +operator-sdk:gen-csv:customresourcedefinitions.resources="Route,route.openshift.io/v1"
// +operator-sdk:gen-csv:customresourcedefinitions.resources="Service,v1"
// +operator-sdk:gen-csv:customresourcedefinitions.resources="KafkaTopic,kafka.strimzi.io/v1beta1"
```

**Review by ricardozanini — CHANGES_REQUESTED** [review]
Looks good, there's the change in the metadata that needs review thou

File: cmd/kogito/command/shared/install_services.go [hunk 3]
```diff
@@ -55,6 +55,9 @@ type ServicesInstallation interface {
 	// InstallDataIndex installs Data Index. If no reference provided, it will install the default instance.
 	// Depends on the Operator, install it first.
 	InstallDataIndex(dataIndex *v1alpha1.KogitoDataIndex) ServicesInstallation
+	// InstallExplainability installs Explainability. If no reference provided, it will install the default instance.
+	// Depends on the Operator, install it first.
+	InstallExplainability(trusty *v1alpha1.KogitoExplainability) ServicesInstallation
```
Comment (sutaakar) [inline cmd/kogito/command/shared/install_services.go:60]:
```suggestion
	InstallExplainability(explainability *v1alpha1.KogitoExplainability) ServicesInstallation
```

File: cmd/kogito/command/shared/install_services.go [hunk 4]
```diff
@@ -149,6 +152,19 @@ func (s *servicesInstallation) InstallDataIndex(dataIndex *v1alpha1.KogitoDataIn
 	return s
 }
 
+func (s *servicesInstallation) InstallExplainability(trusty *v1alpha1.KogitoExplainability) ServicesInstallation {
```
Comment (sutaakar) [inline cmd/kogito/command/shared/install_services.go:155]:
```suggestion
func (s *servicesInstallation) InstallExplainability(explainability *v1alpha1.KogitoExplainability) ServicesInstallation {
```

File: cmd/kogito/command/shared/install_services.go [hunk 5]
```diff
@@ -149,6 +152,19 @@ func (s *servicesInstallation) InstallDataIndex(dataIndex *v1alpha1.KogitoDataIn
...
+	if s.err == nil {
+		s.err = s.installKogitoService(trusty,
```
Comment (sutaakar) [inline cmd/kogito/command/shared/install_services.go:157]:
```suggestion
		s.err = s.installKogitoService(explainability,
```

File: deploy/crds/app.kiegroup.org_v1alpha1_kogitoexplainability_cr.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  # environment variables to set in the runtime container. Example: JAVAOPTS: "-Dquarkus.log.level=DEBUG"
+  # env: {}
+  # number of pods to be deployed
+  replicas: 1
+  # httpPort: 8080
+  # details about the kafka connection
+  kafka:
+    # let's leave this burden to KogitoInfra CR to deploy a new Kafka instance for us
+    useKogitoInfra: false
+  # details about the connected infinispan
```
Comment (sutaakar) [inline deploy/crds/app.kiegroup.org_v1alpha1_kogitoexplainability_cr.yaml:15]: Explainability service doesn't use Infinispan, right?
This line can be removed then,

Comment (r00ta) [reply deploy/crds/app.kiegroup.org_v1alpha1_kogitoexplainability_cr.yaml]: Yep, thanks for the catch, you're right! Removed

File: pkg/controller/kogitoexplainability/kogitoexplainability_controller.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+	err = c.Watch(&source.Kind{Type: &appv1alpha1.KogitoExplainability{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// TODO(user): Modify this to be the types you create that are owned by the primary resource
+	// Watch for changes to secondary resource Pods and requeue the owner KogitoExplainability
+	err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &appv1alpha1.KogitoTrusty{},
```
Comment (sutaakar) [inline pkg/controller/kogitoexplainability/kogitoexplainability_controller.go:73]:
```suggestion
		OwnerType:    &appv1alpha1.KogitoExplainability{},
```

File: test/features/install_explainability.feature [hunk 8]
```diff
@@ -0,0 +1,53 @@
+@explainability
+@kafka
+Feature: Kogito Data Index
```
Comment (sutaakar) [inline test/features/install_explainability.feature:3]:
```suggestion
Feature: Kogito Explainability service
```

File: test/features/install_explainability.feature [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+  Background:
+    Given Namespace is created
+    And Kogito Operator is deployed with Kafka operator
```
Comment (sutaakar) [inline test/features/install_explainability.feature:7]: This line can be moved to `Install Kogito Explainability` scenario to Given part, as the other scenario use also Infinispan operator.

**Comment by sutaakar** [comment]
The code looks ok, taking a look on the BDD test now.
@r00ta Should the test work using latest nightly images?

File: test/steps/kogitoexplainability.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+	return framework.InstallKogitoExplainabilityService(data.Namespace, framework.GetDefaultInstallerType(), explainability)
+}
+
+func (data *Data) kogitoExplainabilityHasPodsRunningWithinMinutes(podNb, timeoutInMin int) error {
+	return framework.WaitForKogitoExplainabilityService(data.Namespace, podNb, timeoutInMin)
+}
+
+type executionsResponse struct {
+	Executions []execution `json:"executions"`
```
Comment (sutaakar) [inline test/steps/kogitoexplainability.go:66]: When I retrieved content of <trusty URL>/executions then the response is:
`{\"total\":1,\"limit\":100,\"offset\":0,\"headers\":[{\"executionId\":\"ac61b2ac-d02d-4349-b82d-72656c1ca8de\",\"executionDate\":\"2020-09-07T13:23:00+0000\",\"executionSucceeded\":true,\"executorName\":null,\"executedModelName\":\"LoanEligibility\",\"executedModelNamespace\":\"https://github.com/kiegroup/kogito-examples/dmn-quarkus-listener-example\",\"executionType\":\"DECISION\"}]}"}
`

So, should the JSON marshalling hint be adjusted?
```suggestion
	Executions []execution `json:"headers"`
```

Comment (r00ta) [reply test/steps/kogitoexplainability.go]: Yep, thank you!

**Comment by r00ta** [comment]
> The code looks ok, taking a look on the BDD test now.
> @r00ta Should the test work using latest nightly images?

Hi @sutaakar , yes, the test should work

**Review by ricardozanini — APPROVED** [review]
Waiting for @sutaakar tests

**Comment by sutaakar** [comment]
@r00ta while running the BDD tests I hit this issue - https://issues.redhat.com/browse/KOGITO-3295

**Comment by r00ta** [comment]
Hi @sutaakar , may I ask you to run the tests again? I should have fixed it. I'm still trying to run the bdd tests locally..

**Comment by sutaakar** [comment]
@r00ta The `externalURI` in KogitoRuntime CR is being overwritten by Kogito operator, so it becomes ineffective.
Sorry for a delay in response.

File: test/features/install_explainability.feature [hunk 11]
```diff
@@ -0,0 +1,54 @@
+@explainability
+@kafka
+Feature: Kogito Explainability service
+
+  Background:
+    Given Namespace is created
+
+  @smoke
+  Scenario: Install Kogito Explainability
```
Comment (sutaakar) [inline test/features/install_explainability.feature:9]:
```suggestion
  Scenario: Install Kogito Explainability
```
```suggestion
  # Disabled as long as https://issues.redhat.com/browse/KOGITO-3176 is not solved
  @disabled
  Scenario: Install Kogito Explainability
```

**Comment by sutaakar** [comment]
@r00ta one thing is still not clear to me:
If I understand it correctly from the description in this PR the user request can trigger several decisions. Every decision is processed by Trusty service which will trigger Explainability request.
Lets say that user sent a 5 requests to Kogito which triggered 10 decisions. Now as a user I would like to see explainability informations about request number 3. How can I pair the explainability results with Kogito request?

**Comment by r00ta** [comment]
> @r00ta one thing is still not clear to me:
> If I understand it correctly from the description in this PR the user request can trigger several decisions. Every decision is processed by Trusty service which will trigger Explainability request.
> Lets say that user sent a 5 requests to Kogito which triggered 10 decisions. Now as a user I would like to see explainability informations about request number 3. How can I pair the explainability results with Kogito request?

Hi @sutaakar , of course there might be multiple decisions taken by Kogito given a particular input, but only one trace event is emitted and received by the trusty service. So it's a 1:1 mapping between the requests of the users and the trace events that are generated. 
On the other side, the explainability services takes one explainability requests and run multiple (mapping 1:N) requests to kogito, but it's only an internal detail since the explainability service combines all those requests so to provide only the explainability result of the original request. 
In short, the user executes a particular decision (let's say it has the `executionId` "my-executionId"), then he can query the trusty service to retrieve all the details and the explainability of that execution. All the `N` requests that the explainability service executes are meaningless for the user, they are only used to calculate the explainability of the execution  "my-executionId"

**Comment by msmagnanijr** [comment]
I'm seeing a lot of:  {"level":"error","T":"2020-09-10T14:25:06.758Z","logger":"jenkins-cucumber-9a1c","msg":"Error while retrieving log of pod 'strimzi-cluster-operator-v0.19.0-5f48687d5d-76xxw': the server has asked for the client to provide credentials ( pods/log strimzi-cluster-operator-v0.19.0-5f48687d5d-76xxw)"}

Jenkins seems to be working properly, and we have a lot of resources. Any idea?

**Comment by sutaakar** [comment]
I haven't seen such issue yet

**Comment by sutaakar** [comment]
In the logs I see:
```

{"level":"error","T":"2020-09-10T14:07:05.008Z","logger":"jenkins-cucumber-c502","msg":"output command: Kogito Build Service successfully installed in the Project jenkins-cucumber-c502.\nCheck the Service status by running 'oc describe kogitobuild/process-timer-quarkus -n jenkins-cucumber-c502'\nThe provided source is a directory, packing files.\nFile(s) found: [/tmp/kogito-examples699273148/process-timer-quarkus/src/main/resources/application.properties /tmp/kogito-examples699273148/process-timer-quarkus/src/main/resources/org/acme/travels/timer-on-task.bpmn /tmp/kogito-examples699273148/process-timer-quarkus/src/main/resources/org/acme/travels/timers.bpmn /tmp/kogito-examples699273148/process-timer-quarkus/src/main/resources/org/acme/travels/timersCycle.bpmn].\nTriggering the new build\nInternal error occurred: unable to upgrade connection: Unauthorized\n"}
```
Seems like the OCP token expired in the middle of the test.

**Comment by r00ta** [comment]
Hi @sutaakar , after fixing some bugs here and there on the operator to run the bdd tests locally I finally managed to run them locally. May I ask you to run them locally as well?

File: test/steps/mappers/create_kogito_service.go [hunk 12]
```diff
@@ -167,6 +168,9 @@ func mapKogitoServiceConfigTableRow(row *TableRow, kogitoService *bddtypes.Kogit
 	case kogitoServiceInfinispanEnableEventsKey:
 		kogitoService.EnableEvents = MustParseEnabledDisabled(getThirdColumn(row))
 
+	case kogitoServiceExternalURIKey:
+		kogitoService.KogitoService.GetStatus().SetExternalURI(getThirdColumn(row))
```
Comment (r00ta) [inline test/steps/mappers/create_kogito_service.go:172]: Hi @sutaakar , I found out that this code does not really set the status/externalURI, so the kogito runtime env variable `kogito.service.url` is not populated and by consequence the test fails. 
Do you have any tip to create a deployment like the following yaml?
```
apiVersion: app.kiegroup.org/v1alpha1
kind: KogitoRuntime
metadata:
  name: dmn-tracing-quarkus
spec:
  replicas: 1
  image:
    domain: quay.io
    namespace: jrota
    name: dmn-tracing-quarkus
    tag: 1.0.6-snapshot
    #  selector:
    # app: dmn-tracing-quarkus 
  kafka:
    externalURI: external-kafka-kafka-bootstrap:9092
status:
  externalURI: "http://dmn-tracing-quarkus:8080"
  conditions:
    - lastTransitionTime: "1985-04-12T23:20:50.52Z"
      lastUpdateTime: "1985-04-12T23:20:50.52Z"
      message: ""
      reason: ""
      status: ""
      type: ""
```

Comment (sutaakar) [reply test/steps/mappers/create_kogito_service.go:172]: Hi @r00ta IMHO we shouldn't set status fields, as this fields are used by Kogito operator to reflect current component status.
Could be `kogito.service.url` set as a runtime env variable directly in KogitoRuntime specs?

Comment (r00ta) [reply test/steps/mappers/create_kogito_service.go:172]: Allright.. I found out that portion of is ignored by the cli (which is used to install the services during the tests)..

Comment (r00ta) [reply test/steps/mappers/create_kogito_service.go:172]: Hi @sutaakar , I found out that the operator is anyway overwriting that env variable with the value in `kogitoService.KogitoService.GetStatus().GetExternalURI()`, I'm going to modify that code so that the operator sets that value if and only if `kogitoService.KogitoService.GetStatus().GetExternalURI() != ""`

**Comment by r00ta** [comment]
Hi @sutaakar @ricardozanini , I've been finally able to run the bdd with success. Now the tests should be running fine. I had to modify a little bit the test framework since it was not working on `minikube`. I'm not sure it depends on my local configuration, but for example the function `GetIngressURI` has hardcoded the target port `80`, but this is not the case usually since the default port of the services is `8080`. 
Another thing is `ExposeServiceOnKubernetes`: it creates an ingress resource for every service, but they are overlapping at the path `/`. I quickly hacked it locally so to make it work. So it might be that nobody uses `kubernetes` to run the bdd tests or that my env is messed up.. If you confirm this is an issue for everybody, let me know and I'll create a ticket. 

Could you test it locally as well? Everything should be working now. Thank you very much guys : )

**Comment by ricardozanini** [comment]
Waiting for #558 CI to finish

File: cmd/kogito/command/install/explainability.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+func (i *installExplainabilityCommand) RegisterHook() {
+	i.command = &cobra.Command{
+		Use:     "explainability [flags]",
+		Short:   "Installs the Kogito Explainability Service in the given Project",
+		Example: "explainability -p my-project",
+		Long: `'install explainability' will deploy the Explainability service to provide analysis on the decisions that have been taken by a kogito runtime application.
+
+If kafka-url is provided, it will be used to connect to the external Kafka server that is deployed in other project or infrastructure.
+If kafka-instance is provided instead, the value will be used as the Strimzi Kafka instance name to locate the Kafka server deployed in the Explainability service's project.
```
Comment (spolti) [inline cmd/kogito/command/install/explainability.go:72]: to locate the deployed Kafka..

**Review by spolti — CHANGES_REQUESTED** [review]
just a little detail on command help.

**Review by spolti — APPROVED** [review]

**Comment by ricardozanini** [comment]
@r00ta unit tests failing. I'll run the jenkins test to see if everything is ok there

/jenkins test

**Comment by ricardozanini** [comment]
I retriggered manually the CI check, the first one ended with an error in the OCI.
Ongoing: https://jenkins-kogito-tools.apps.kogito.automation.rhmw.io/blue/organizations/jenkins/Kogito%2Fkogito-operator-pr-check/detail/kogito-operator-pr-check/262/pipeline/

**Comment by ricardozanini** [comment]
Explainability test errors:

```xml
<testsuite name="Kogito Explainability service" tests="1" skipped="0" failures="1" errors="0" time="1043.189412945">

    <testcase name="Explainability retrieves explainability requests events using external Kafka" status="failed" time="1043.189412945">

      <failure message="Step Explainability result is available in the Trusty service within 3 minutes: Timeout waiting for GET request on path executions/decisions/e7885d9e-6c2f-46e1-bdba-c80b2b458c10/explanations/saliencies to return response content &#39;SUCCEEDED&#39;"></failure>

    </testcase>

  </testsuite>
```

Also Keycloak:

```xml
<testsuite name="Kogito integration with Keycloak" tests="2" skipped="0" failures="1" errors="0" time="1251.224082765">

    <testcase name="Install Kogito Data Index with Keycloak security" status="passed" time="644.797874048"></testcase>

    <testcase name="Install Kogito Jobs Service with Keycloak security" status="failed" time="1251.206813469">

      <failure message="Step Keycloak instance with realm &#34;kogito-realm&#34; and client &#34;kogito-jobs-service&#34; is deployed: Timeout waiting for Pods with label name &#39;app&#39; and value &#39;keycloak&#39; available and running"></failure>

      <error message="Step Keycloak user &#34;my-user&#34; with password &#34;my-password&#34; is deployed" type="skipped"></error>

      <error message="Step Install Kogito Jobs Service with 1 replicas with configuration:" type="skipped"></error>

      <error message="Step Kogito Jobs Service has 1 pods running within 10 minutes" type="skipped"></error>

      <error message="Step HTTP GET request on service &#34;jobs-service&#34; with path &#34;jobs&#34; is forbidden within 1 minutes" type="skipped"></error>

      <error message="Step Stores access token for user &#34;my-user&#34; and password &#34;my-password&#34; on realm &#34;kogito-realm&#34; and client &#34;kogito-jobs-service&#34; into variable &#34;my-user-token&#34;" type="skipped"></error>

      <error message="Step HTTP POST request on service &#34;jobs-service&#34; using access token &#34;{my-user-token}&#34; is successful within 2 minutes with path &#34;jobs&#34; and body:" type="skipped"></error>

      <error message="Step HTTP GET request on service &#34;jobs-service&#34; using access token &#34;{my-user-token}&#34; with path &#34;jobs/1&#34; is successful within 1 minutes" type="skipped"></error>

    </testcase>

  </testsuite>
```

**Comment by ricardozanini** [comment]
@r00ta you need to rebase.

**Comment by r00ta** [comment]
cherry pick on 0.15.x is here https://github.com/kiegroup/kogito-cloud-operator/pull/568

**Review by Sgitario — COMMENTED** [review]
Hi @r00ta , can you add the new KogitoExplainabilityList custom resource in [this line](https://github.com/kiegroup/kogito-cloud-operator/blob/3fa1d80556ac5d89f26b4be17145196576863f19/test/steps/data.go#L112) as well? This is really useful to debug purposes after running the scenarios. Thanks!

**Comment by r00ta** [comment]
Hi @Sgitario , done

**Comment by ricardozanini** [comment]
It seems that QE needs more time to verify this feature, @r00ta we will push to 0.16 and I will start the release of 0.15 version. Let me know what you think.

**Comment by r00ta** [comment]
> It seems that QE needs more time to verify this feature, @r00ta we will push to 0.16 and I will start the release of 0.15 version. Let me know what you think.

Hi @ricardozanini , the operator is not providing any additional feature: it's just about provisioning of resources at the end. The features were approved on kogito-runtimes and kogito-apps in the last weeks. Is there anything in particular we have to test in this PR?

**Comment by ricardozanini** [comment]
@r00ta I'm not worried about the deployment itself, but I saw @Sgitario comment in the 0.15 branch, it seems that needs some other fixes. Let's finish this discussion here: https://kie.zulipchat.com/#narrow/stream/236603-kogito-ci/topic/.5B0.2E15.2Ex.5D.20Release.20Pipeline
